Subscription management implementation

fselmo opened the pull request:
### What was wrong?

Related to Issue #3397

Working with many subscriptions is not ideal at the moment. In web3.py v7, we added significant support for `eth_subscribe` via the new `PersistentConnectionProvider` implementation. Though we've added support with parity to how `websockets` works, this still only left us with a raw API to handle messages as they come through the socket. We have JSON-RPC specifications of how subscriptions should be formed and the client returns a unique `id` per subscription. Since `PersistentConenctionProvider` classes are fully asynchronous, as they should be, we should implement an API that can leverage all of this in order to handle subscriptions asynchronously as they come in.

### How was it fixed?

- Implement a subscription manager class to manage all subscriptions under-the-hood. This manager can be accessed directly to deal with subscriptions as objects, rather than free-floating ids that are tracked. Each `EthSubscription` class can, and should be encouraged to, have a `handler` function passed to it so that it can properly handle the subscription as it comes in, rather than having the burden of listening and parsing based on message `id` be on the user.
- `w3.eth.subscribe` can now also accept a `handler` kwarg since all subscriptions are managed by the manager under-the-hood. It can also accept an `event` for "logs" subscriptions so that it may be easily parsed by the handler.
- Subscriptions can have descriptive labels for clearer log messages, if desired, instead of relying on an abstract subscription `id` to parse logs.
- The manager counts how many total handler calls have been issued and each subscription counts how many times its handler has been called.

---

In a separate PR we should add a significant docs overhaul for clearer communication on `PersistentConnectionProvider` usage as well as subscription management usage

### Todo:

- [x] Clean up commit history
- [x] Early documentation before a significant docs overhaul to highlight this usage more appropriately
- [x] Add tests
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![1000029633](https://github.com/user-attachments/assets/3d8d09ae-14d6-46c6-8129-44607e0d51c8)

fselmo committed:
feat: Add a subscription manager to manage subscriptions

- Create a SubscriptionManager class to manage subscriptions.
- Implement subscribing to one or many subscriptions.
- Implement unsubscribing from one or many subscriptions.

fselmo committed:
Document changes related to subscription management.

fselmo committed:
Better organize provider property assignment and superclass calls:

- Explicitly set instance propertiece in the constructor, rather than
  the class.
- Order property assignments in the constructor so that the superclass
  is initialized first, with explicit mention when it's useful to
  initialize a particular property on the subclass first.

fselmo committed:
More sx manager additions:

- When subscribing, create typed subscriptions based on the
  subscription params.

- Add ``get_by_id()`` and ``get_by_label()`` methods to the
  ``SubscriptionManager`` class.

- Allow event to be passed to logs subscriptions for easier
  processing when the messages are received.

- Add ability to count handler calls on both the manager and on a per-subscription basis

- Add tests for subscription manager; tighten up typing.

- Fix some types:
  - ``process_log`` expects a ``LogReceipt``, not a hex.
  - ``FormattedEthSubscriptionResponse`` should also include
    ``HexBytes`` for pending tx hashes (full_transactions=False).

- Create live integration tests for ``eth_subscribe`` for "newHeads", "newPendingTransactions", "logs". Syncing still has to be stubbed out.

fselmo commented:
This is ready for a review :)

fselmo committed:
Fix old test that was not working as expected.

fselmo committed:
Context-based approach to subscription handlers args management:

- Introduce a ``handler_context`` to expose the w3, subscription, and result, as well as any user-defined context args.

- Add ``handler_context`` to subscriptions that allow for defining custom
  arguments in the subscription handler implementation, available via
  the context.

- Instead of using transient flags, use the handler context directly
  to create handler tests and assert that the handlers were called
  and the ``passed`` flag was set to ``True``.

- ``event`` for logs subscriptions should be handled by the handler context.

- Changes from comments on #3554; update docs.

- newsfragment for #3554

fselmo commented:
I moved to a context-based args management for subscription handlers in order to facilitate custom args to be used within handlers. This is back on a freeze for reviews.

reedsa commented on web3/_utils/module_testing/module_testing_utils.py:
```diff
@@ -106,7 +106,6 @@ def assert_contains_log(
     emitter_contract_address: ChecksumAddress,
     txn_hash_with_log: HexStr,
 ) -> None:
-    assert len(result) == 1
```
Removing this may cause an index out of range when retrieving the first result. Not sure of the intent here but it's fine with me if we can assume there's a result and potentially have more than one.

reedsa commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,143 @@
+        Used to unsubscribe from a subscription that is being managed by the
+        subscription manager.
+
+        :param subscription: The subscription to unsubscribe from.
```
Might include the type in the docs as well with `:type subscription: EthSubscription[Any]`

reedsa commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,143 @@
+        unsubscribed from or, if `run_forever` is set to `True`, it will run
+        indefinitely.
+
+        :param run_forever: If `True`, the method will run indefinitely.
```
Type for this could be included `:type run_forever: bool`

reedsa commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,143 @@
+        if all(unsubscribed):
+            self.logger.info("Successfully unsubscribed from all subscriptions.")
+        else:
+            self.logger.warning(
```
Could be useful to include which subscriptions are still active. Is it possible to `try` for each `self.unsubscribe` call and bubble up any exceptions that occur?

reedsa reviewed (commented):
Left a few ideas for consideration but didn't see anything that sticks out as a showstopper. Nice work!

wolovim reviewed (approved):
looking primarily at the public API -- this feels like a friendlier way to manage even one subscription. context-aware handlers and subscription type classes are great devex improvements. will be important to have a good reference guide in the docs to surface whats hidden behind context (follow-on 🤝). nice initiative!

kclowes commented on docs/internals.rst:
```diff
@@ -430,10 +430,122 @@ request. The initial ``eth_subscribe`` request expects only one response, the
+of a few ways.
+
+The recommended way to handle one-to-many responses is to use the subscription manager
+API. The subscription manager API is a public API on the ``AsyncWeb3`` class, when
```
nit:
```suggestion
API. The subscription manager API is a public API on the ``AsyncWeb3`` class. When
```

kclowes commented on docs/internals.rst:
```diff
@@ -430,10 +430,122 @@ request. The initial ``eth_subscribe`` request expects only one response, the
+.. code-block:: python
+
+    >>> async def new_heads_handler(
+    ...     context: NewHeadsSubscriptionContext,
```
is this `handler_context` now?

fselmo replied:
The object is, but you can give it whatever name you want, it's not tied to the name `handler_context`. I can change it here to be more clear though.

kclowes replied:
Oh, I misread. No need to change unless you've already done it

kclowes commented on tests/core/providers/test_websocket_provider.py:
```diff
@@ -270,11 +270,13 @@ async def test_listen_event_awaits_msg_processing_when_subscription_queue_is_ful
     async_w3.provider._request_processor._request_information_cache.cache(
-        "", sub_request_information
+        # cache key is the result of `generate_cache_key` with the sub_id as the arg
+        "0138b5d63d66121d8a6e680d23720fa7",
```
may be worth just calculating that here 🤷 
```suggestion
        generate_cache_key(sub_id)
```

fselmo replied:
A little more overhead for a test with an expected result, but makes it clearer 👍🏼

kclowes commented on web3/eth/async_eth.py:
```diff
@@ -714,17 +719,23 @@ async def subscribe(
             ]
         ] = None,
+        handler: Optional[EthSubscriptionHandler] = None,
+        handler_context: Dict[str, Any] = None,
```
nit: 
```suggestion
        handler_context: Optional[Dict[str, Any]] = None,
```

kclowes commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,151 @@
+            if len(self.subscriptions) > 0:
+                self.logger.warning(
+                    "Failed to unsubscribe from all subscriptions. Some subscriptions "
+                    "are still active."
```
Do you think there is a benefit in keeping both the debug log and the warning log here?

fselmo replied:
Debug includes the warning as well, but if you only set the log level to warning (warning and errors only) you will still see this log. Happy to remove though, thoughts?

fselmo replied:
Maybe putting the debug content (the subscriptions that are still present) into the warning? 🤔 

edit: I went ahead and combined them into the warning. Let me know if you have another idea there.

kclowes commented on web3/providers/persistent/websocket.py:
```diff
@@ -59,18 +59,19 @@ class WebSocketProvider(PersistentConnectionProvider):
         self.endpoint_uri = (
             URI(endpoint_uri) if endpoint_uri is not None else get_default_endpoint()
         )
+        super().__init__(**kwargs)
```
It looks like we do a fair amount of moving around the `super().__init__...` call throughout this PR - I'm curious about why we need to do that now, but not before?

fselmo replied:
I think it's cleaner to understand what arguments need to be instantiated before the `super.__init__()` call and describe them inline. There's some reliance on a few arguments needing to exist beforehand so I moved everything else after this init to make that distinction more clear for a reader.

I dedicated an isolated [commit](https://github.com/ethereum/web3.py/pull/3554/commits/4c313eac862d9d0c6eceafbe42f9228357b7c7b5) to these changes to make it a bit more clear.

kclowes commented on web3/utils/subscriptions.py:
```diff
@@ -0,0 +1,285 @@
+    ) -> None:
+        sx = context.subscription
+        sx.handler_call_count += 1
+        sx.manager.total_handler_calls += 1
```
It's very possible I missed this, but other than testing, where does this get used?

fselmo replied:
It's simply for keeping track. As a user I'd like to know these stats from a subscription manager if it's doing some background work that I can't see. If you think we should remove it, I can do so. Thoughts?

kclowes commented on web3/_utils/module_testing/persistent_connection_provider.py:
```diff
@@ -8,245 +11,222 @@
+) -> None:
+    w3 = context.async_w3
+    sx = context.subscription
+    assert w3 is not None
```
Maybe: 
```suggestion
    assert isinstance(w3, AsyncWeb3)
```

kclowes reviewed (approved):
Sweet! I didn't see anything major. This doesn't break the existing API, right? And then feel free to take or leave, but maybe `sub` or `subn` for an abbreviation of subscription instead of `sx` 😆 ?

fselmo committed:
Address remaining comments from PR #3554:

- ``sx`` -> ``sub``.
- Some clarifying changes on type hints and arg names in tests.
- Docs update to clarify handlers. This whole area of the docs will get
  a lot of attention and refactoring in a future PR. But at least the
  current docs are more descriptive.

fselmo commented:
> This doesn't break the existing API, right?

@kclowes, always good to ask 😅. The relevant API changes only [introduce more customization](https://github.com/ethereum/web3.py/pull/3554/files#diff-745353d44729acfd441365c5fe774d4b66affab2319ef3e1ab683dbed1885e38R722-R738). This shouldn't break the API. And if no handlers are passed in (new feature), you can still listen to free-flowing messages directly from the socket.

The typing in the socket stream was already `Optional` meaning a `None` could technically be returned as a "message". If a message has a handler a `None` is returned so that the iterator can move on to the next iteration. This is the only friction I see this PR introducing. I didn't see a clean way to move a users `async for message in socket` to the next message, if it was handled in a handler, without yielding a `None` but perhaps we can investigate that a bit more.

Thoughts on that?

fselmo committed:
Separate handler subscriptions into their own queue:

- Make use of a ``TaskReliant`` queue that is specific for handler
  subscriptions. Put a ``SubscriptionProcessingFinished`` exceptions
  in the queue to signal when all handler subscriptions have been
  unsubscribed from.

- Add testing around running the ``handler_subscriptions`` method while
  listening for un-handled subscriptions via the socket.

- Create a container to manage all subscriptions, since we store subs
  in more than one way: by ``id``, by ``label``, etc...

- The ``SubscriptionContainer`` class has an object reference, so we changes
  share this with the provider's ``RequestProcessor`` class. The
  request_processor can use the container to make decisions about which queue
  to store the subscription in, based on the presence or absence of a
  ``handler``.

kclowes reviewed (approved):
I like this! I didn't have time to go through with a fine tooth comb, but if you feel good about it and the tests, don't let me block your merge 👍

fselmo committed:
Handle flaky tests a bit better:

Properly close tasks related to each test before ending tests:

- Properly cancel and close all tasks for persistent connection provider tests
- Clear caches as a cleanup for all persistent connection provider tests

Increase dev period to allow `replace_transaction` tests more time:

- Re-running these tests with flaky does not help, hey will end up failing
  every try. Increase the dev period to allow `replace_transaction` tests
  more time and remove the flaky decorator.

- Lower request timeout for tests. No test should need more than 10s.

fselmo commented:
> I like this! I didn't have time to go through with a fine tooth comb, but if you feel good about it and the tests, don't let me block your merge 👍

It has been pretty battle tested but it's causing some irrelevant flaky replace transaction tests to fail pretty consistently, just on CI. I checked against `main` and it is specific to this branch. I'll wait a little longer to debug it and try to figure out what's causing it. The frustrating thing is that locally it isn't reproducible (these flaky tests haven't been reproducible locally all along). So I might push a commit on top of these until I can sniff it out and try to assess, roll the commit back, and handle appropriately.

My suspicion is it's not directly related to this PR. Ultimately those tests need to be properly fixed, so that might be what I end up doing in a separate PR.

Either way I'll be sure to preserve the current state of the new code (no force pushing on top of the previous commits) so that nothing that you all have reviewed changes.

fselmo committed:
[Persistent Connection Providers] Improve request info management:

- Instead of guessing what the request `id` will be and having to bump
  this cache key for internal requests, split persistent Connection
  provider calls into send and recv. Use the build request after sending
  to get the `id` and store the request information for when the response
  is received.

- Add request caching support and request mocker support for these new
  methods. Use the `send_request` and `recv_for_request` methods on the
  `Module` class so we properly handle the request information for
  processing the response.

fselmo commented on web3/manager.py:
```diff
@@ -513,20 +512,41 @@ async def socket_request(
+        )
         self.logger.debug(
-            "Sending request to open socket connection: "
+            "Making request to open socket connection and waiting for response: "
```
Note: This should be put back to its original message... got mixed with make_request somewhere along the way

kclowes commented on web3/utils/subscriptions.py:
```diff
@@ -0,0 +1,285 @@
+    return wrapped_handler
+
+
+class EthSubscription(Generic[TSubscriptionResult]):
```
This is pretty nitty, so feel free to take or leave, but what do you think about just calling this `Subscription`?

fselmo replied:
I'm into it 👍🏼

fselmo replied:
oo I think I realized why I did this. We already had `EthSubscriptionParams` and `FormattedEthSubscriptionResponse` as types in web3.types. I added `EthSubscription` (class) and `EthSubscriptionResult` (type) in this PR. I'd still lean toward keeping it `Eth...` but I can be convinced that the class can be a different name. Thoughts there?

kclowes replied:
I'm good with keeping it as `Eth` if that's the convention! 👍

kclowes commented on web3/utils/subscriptions.py:
```diff
@@ -0,0 +1,285 @@
+        handler: Optional[PendingTxSubscriptionHandler] = None,
+        handler_context: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.full_transactions = full_transactions
```
nit: are you using this anywhere?

fselmo replied:
No, this was a choice to set the properties on the class so it's available. May also be useful when handling subscriptions in the handlers.

kclowes replied:
Gotcha, that makes sense. Maybe a comment (or test if there isn't one?) so it doesn't get accidentally removed along the way?

fselmo replied:
I did miss it in the properties test. Logs checks the `address` and `topics` properties as well. I added this in the pending tx properties core test 👍🏼

reedsa commented:
Your docs are great and I wonder if they would be worth referencing in the [Eth.subscribe](https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.subscribe) section and/or the Providers pages? Might help guide more advanced usage there.

fselmo commented:
> Your docs are great and I wonder if they would be worth referencing in the [Eth.subscribe](https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.subscribe) section and/or the Providers pages? Might help guide more advanced usage there.

Good thought. We're going to do a big overhaul of these docs so this is just some basic stuff on usage for now. Don't want to put too much into this since it may change quite a bit. Thoughts?

kclowes commented on web3/providers/persistent/persistent.py:
```diff
@@ -72,9 +86,63 @@ def __init__(
+
+            self._send_func_cache = (cache_key, send_function)
+
+        return self._send_func_cache[1]
```
nit: maybe just `return send_function` here?

fselmo replied:
We need to return the cached function here. We don't always build it when we call `send_func` from the manager so this pulls it from the cache instead. It's how the `request_func` is set up as well. Thoughts?

kclowes replied:
Ahh, I see. That makes sense to me. Clever :)

kclowes commented on web3/providers/persistent/persistent.py:
```diff
@@ -72,9 +86,63 @@ def __init__(
+
+            self._recv_func_cache = (cache_key, recv_function)
+
+        return self._recv_func_cache[1]
```
same nit here, maybe:
```suggestion
        return recv_function
```

fselmo replied:
See other comment.

kclowes commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,274 @@
+    )
+
+
+class SubscriptionContainer:
```
more nits: Do you think this `SubscriptionContainer` class deserves its own file?

fselmo replied:
I could go either way. Do you? 😆 I don't think this file is too large but happy to abstract it out. Thoughts?

kclowes replied:
I could go either way too. I think I'd slightly prefer it, but not a strong opinion and we can always do it later if it gets unwieldy :)

fselmo replied:
No prob, I'll add it in here 👍🏼

kclowes commented on web3/providers/persistent/subscription_manager.py:
```diff
@@ -0,0 +1,274 @@
+    def __iter__(self) -> Iterator[EthSubscription[Any]]:
+        return iter(self.subscriptions)
+
+    def add_subscription(self, subscription: EthSubscription[Any]) -> None:
```
Do you think it's worth adding a tighter type here rather than the `Any`? Assuming I'm reading this right, there are well-defined params we allow here, right?

fselmo replied:
If I remember correctly, this allows us to still pass in and handle unknown subscriptions to our library, since we only define the execution-spec'd subscriptions. Alchemy has some [custom subs](https://docs.alchemy.com/reference/subscription-api) they define, for example. I'll see if that was the reason and if so, I think we can leave this be since we define tighter types on the defined subscription classes themselves where we know the results.

kclowes replied:
Cool, I'm good with leaving it!

kclowes commented on web3/_utils/module_testing/eth_module.py:
```diff
@@ -2391,10 +2388,6 @@ async def test_async_eth_replace_transaction_gas_price_too_low(
         with pytest.raises(Web3ValueError):
             await async_w3.eth.replace_transaction(txn_hash, txn_params)
 
-    @flaky_with_xfail_on_exception(
```
🎉

kclowes reviewed (approved):
This looks great! I had a some nits throughout, but feel free to take or leave 'em!

reedsa commented:
> > Your docs are great and I wonder if they would be worth referencing in the [Eth.subscribe](https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.subscribe) section and/or the Providers pages? Might help guide more advanced usage there.
> 
> Good thought. We're going to do a big overhaul of these docs so this is just some basic stuff on usage for now. Don't want to put too much into this since it may change quite a bit. Thoughts?

SGTM! Looks good to go from my perspective.

fselmo committed:
Add signal handlers for graceful disconnects:

- On ``SIGTERM`` and  ``SIGNINT`` (KeyboardInterrupt), disconnect
  gracefully from the provider.

- KeyboardInterrupt used to hang for a while because of the open
  connection. This makes it so we no longer wait the whole
  connection_timeout and it raises the KeyboardInterrupt exception
  immediately. Leave the responsibility of handling the
  KeyboardInterrupt to the user.

fselmo committed:
More changes from comments on PR #3554:

- Test `full_transactions` property of `PendingTxSubscription`
- Abstract `SubscriptionContainer` to its own file
